File: LeetCode Daily/01143. Longest Common Subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int n=text1.length(),m=text2.length();
        int dp[n+1][m+1];
        memset(dp,0,sizeof(dp));
        for(int i=1;i<=n;i++){
            for(int j=1;j<=m;j++){
                if(text1[i-1]==text2[j-1])dp[i][j]=dp[i-1][j-1]+1;
                else dp[i][j]=max(dp[i][j-1],dp[i-1][j]);
            }
        }
        return dp[n][m];      
    }

    
};

File: LeetCode Daily/01143. Longest Common Subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int[][] memo = new int[text1.length()][text2.length()];
        for (int[] row : memo) {
            Arrays.fill(row, -1);
        }
        return dfs(text1, text2, 0, 0, memo);       
    }

    private int dfs(String text1, String text2, int i, int j, int[][] memo) {
        if (i == text1.length() || j == text2.length()) {
            return 0;
        }
        if (memo[i][j] != -1) {
            return memo[i][j];
        }
        int res = 0;
        if (text1.charAt(i) == text2.charAt(j)) {
            res = Math.max(res, 1 + dfs(text1, text2, i + 1, j + 1, memo));
        } else {
            res = Math.max(res, dfs(text1, text2, i + 1, j, memo));
            res = Math.max(res, dfs(text1, text2, i, j + 1, memo));
        }
        memo[i][j] = res;
        return res;
    }
}